src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,241 @@
+            (tfAllOrNothing | tfOnlyOne | tfUntilFailure | tfIndependent)) != 1)
+    {
+        JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                            << "too many flags.";
+        return temINVALID_FLAG;
+    }
+
+    auto const& rawTxns = ctx.tx.getFieldArray(sfRawTransactions);
+    if (rawTxns.size() <= 1)
+    {
+        JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                            << "txns array must have at least 2 entries.";
+        return temARRAY_EMPTY;
+    }
+
+    if (rawTxns.size() > maxBatchTxCount)
+    {
+        JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                            << "txns array exceeds 8 entries.";
+        return temARRAY_TOO_LARGE;
+    }
+
+    // Validation Inner Batch Txns
+    std::unordered_set<AccountID> requiredSigners;
+    std::unordered_set<uint256> uniqueHashes;
+    for (STObject rb : rawTxns)
+    {
+        STTx const stx = STTx{std::move(rb)};
+        auto const hash = stx.getTransactionID();
+        if (!uniqueHashes.emplace(hash).second)
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                                << "duplicate TxID found."
+                                << "txID: " << hash;
+            return temMALFORMED;
+        }
+
+        if (stx.getFieldU16(sfTransactionType) == ttBATCH)
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                                << "batch cannot have an inner batch txn."
+                                << "txID: " << hash;
+            return temINVALID_INNER_BATCH;
+        }
+
+        if (stx.isFieldPresent(sfTxnSignature) ||
+            stx.isFieldPresent(sfSigners) || !stx.getSigningPubKey().empty())
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                                << "inner txn cannot include TxnSignature and "
+                                   "Signers and SigningPubKey must be empty."
+                                << "txID: " << hash;
+            return temINVALID_INNER_BATCH;
+        }
+
+        auto const innerAccount = stx.getAccountID(sfAccount);
+        if (auto const preflightResult = ripple::preflight(
+                ctx.app, ctx.rules, parentBatchId, stx, tapBATCH, ctx.j);
+            preflightResult.ter != tesSUCCESS)
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]:"
+                                << "inner txn preflight failed."
+                                << "txID: " << hash;
+            return temINVALID_INNER_BATCH;
+        }
+
+        // If the inner account is the same as the outer account, do not add the
+        // inner account to the required signers set.
+        if (innerAccount != outerAccount)
+            requiredSigners.insert(innerAccount);
+    }
+
+    if (auto const ret = preflight2(ctx); !isTesSuccess(ret))
+        return ret;
```

Comment:
Can we get a test covering this line?

Reply:
fixed: https://github.com/XRPLF/rippled/pull/5060/commits/94b403613bdf32afbc6bd7d1fe54ce1b7afc25ad

Reply:
This is not possible to check. This function calls `checkValidity` which is already called by the time it gets to preflight2. So I can test checkValidity but it will not capture this test. I think thats why the error is marked `temINVALID` and not something more related to the signature check.

Reply:
Can we escape it from codecov then?